From 7d2ddc97287a9034afa0924f0006aa97af30d78e Mon Sep 17 00:00:00 2001
From: Stefan Agner <stefan.agner@toradex.com>
Date: Mon, 14 Nov 2016 18:25:16 -0800
Subject: [PATCH 1/2] apalis_imx6: use SDP if recovery mode is enabled

In U-Boot, in case recovery mode is used, set bootcmd to start serial
download protocol (SDP). This allows to download complete images such
as Toradex Easy Installer over USB SDP.

Signed-off-by: Stefan Agner <stefan.agner@toradex.com>
---
 board/toradex/apalis_imx6/apalis_imx6.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/board/toradex/apalis_imx6/apalis_imx6.c b/board/toradex/apalis_imx6/apalis_imx6.c
index 93e777d6ce..6a7d26cb4d 100644
--- a/board/toradex/apalis_imx6/apalis_imx6.c
+++ b/board/toradex/apalis_imx6/apalis_imx6.c
@@ -847,6 +847,9 @@ int board_init(void)
 #ifdef CONFIG_BOARD_LATE_INIT
 int board_late_init(void)
 {
+	struct src *src = (struct src *)SRC_BASE_ADDR;
+	unsigned int gpr10_boot = readl(&src->gpr10) & (1 << 28);
+	unsigned int gpr9 = readl(&src->gpr9);
 #if defined(CONFIG_REVISION_TAG) && \
     defined(CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG)
 	char env_str[256];
@@ -876,6 +879,13 @@ int board_late_init(void)
 #endif /* CONFIG_TDX_APALIS_IMX6_V1_0 */
 #endif /* CONFIG_REVISION_TAG */
 
+	if (((readl(&src->sbmr2) >> 24) & 0x03) == 0x01 ||
+		(gpr10_boot && (gpr9 == 0x1))) {
+		printf("Serial Downloader recovery mode, using sdp command\n");
+		setenv("bootdelay", "0");
+		setenv("bootcmd", "sdp 0");
+	}
+
 	return 0;
 }
 #endif /* CONFIG_BOARD_LATE_INIT */
-- 
2.11.0

